cpp/src/arrow/array/array_nested.cc

```diff
@@ -282,18 +281,24 @@ Result<std::shared_ptr<Array>> FlattenListViewArray(const ListViewArrayT& list_v
     }
   }
 
-  const auto* validity = list_view_array.data()->template GetValues<uint8_t>(0, 0);
   const auto* offsets = list_view_array.data()->template GetValues<offset_type>(1);
   const auto* sizes = list_view_array.data()->template GetValues<offset_type>(2);
 
-  auto is_null_or_empty = [&](int64_t i) {
-    if constexpr (HasNulls) {
+  std::function<bool(int64_t)> is_null_or_empty;
```

Comment:
This is complicated. Instead, how about just:
```c++
  auto is_null_or_empty = [&](int64_t i) {
    if (HasNulls && !bit_util::GetBit(validity, list_view_array_offset + i)) {
      return true;
    }
    return sizes[i] == 0;
  };
```

Reply:
+1.

The compiler will constant-fold it when the lambda gets inlined. Putting it in `std::function` would prevent inlining.